src/components/WebCLI/WebCLI.js

```diff
@@ -0,0 +1,153 @@
+import React, { useCallback, useEffect, useRef, useState } from "react";
+
+import WebCLIOutput from "./Output.js";
+
+import "./_webcli.scss";
+
+const DEFAULT_PLACEHOLDER = "enter command";
+
+const generateAddress = (controllerWSHost, modelUUID) => {
+  if (!controllerWSHost || !modelUUID) {
+    return null;
+  }
+  return `wss://${controllerWSHost}/model/${modelUUID}/commands`;
+};
+
+const WebCLI = ({ controllerWSHost, credentials, modelUUID }) => {
+  const [connection, setConnection] = useState(null);
+  const [placeholder, setPlaceholder] = useState(DEFAULT_PLACEHOLDER);
+  const [shouldShowHelp, setShouldShowHelp] = useState(false);
+  const inputRef = useRef();
+  const wsMessageStore = useRef();
+  let [output, setOutput] = useState("");
+
+  const setDisconnectedPlaceholder = () => {
+    setPlaceholder("no web cli backend available");
+  };
+
+  const setConnectedPlaceholder = () => {
+    setPlaceholder(DEFAULT_PLACEHOLDER);
+  };
+
+  /*
+    The messageBuffer is required because the websocket returns messages much
+    faster than React wants to update the component. Doing this allows us to
+    store the messages in a buffer and then set the output every cycle.
+  */
+  const messageBuffer = useCallback(
+    (message) => {
+      wsMessageStore.current = wsMessageStore.current + message;
+      setTimeout(() => {
```

Comment:
Doesn't this setTimeout require a timeout value?

Reply:
When you use setTimeout without a time it effectively pops it to the top of the stack to be executed on the next loop.